### PR TITLE
Refactored audio and voice support

### PR DESCRIPTION
**Pull request by kaikreuzer**
- Created audio API in smarthome.core
- Introduced smarthome.core.voice for TTS, STT and HLI APIs
- Introduced smarthome.io.javasound for audio implementations based on Java Sound API
- Introduced smarthome.voice.mactts as a Mac specific TTS implementation

Also-By: Harald Kuhn <harald.kuhn@gmail.com>
Also-By: Kelly Davis <kdavis@mozilla.com>
Also-By: Tilman Kamp <tilmankamp@posteo.de>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
This PR contains all the work that was done for #1081, #1028 and #1021.

I have tried to clean it all up, get the tests working, have stuff removed that is not possible to add to ESH due to the license (MaryTTS, PocketSphinx, etc.).

@hkuhn42 Would be great to have you reviewing this or at least that you familiarize yourself with it and see how/if you can further contribute to it (Kelly&Co unfortunately stopped working on it).

I shuffled around the interfaces a bit and decided to put the voice stuff now directly under core.voice and have the audio interfaces in the core bundle itself, so that issues like https://github.com/eclipse/smarthome/issues/1200 can be easily implemented without any optional dependencies.
The only other thing that is probably new to you is the AudioStream interface - I introduced this as a handle of a stream, while keeping AudioSource as an interface for OSGi services. When implementing services, I noticed that it makes much more sense this way.

**Comment by kaikreuzer** [comment]
@maggu2810 FTR: This PR is more than 1000 lines and has contributions from others, so we will require a CQ for it. Once you and @hkuhn42 have given a lgtm, I will create a CQ for it.

**Comment by maggu2810** [comment]
Travic CI failed:

```text
[ERROR] Internal error: org.eclipse.tycho.core.osgitools.OsgiManifestParserException: Exception parsing OSGi MANIFEST /home/travis/build/eclipse/smarthome/bundles/core/org.eclipse.smarthome.core.voice/META-INF/MANIFEST.MF: Manifest file not found -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: org.eclipse.tycho.core.osgitools.OsgiManifestParserException: Exception parsing OSGi MANIFEST /home/travis/build/eclipse/smarthome/bundles/core/org.eclipse.smarthome.core.voice/META-INF/MANIFEST.MF: Manifest file not found
```

**Comment by kaikreuzer** [comment]
Weird, somehow all the new MANIFESTs are not shown (and added) in git...
Give me a moment.

**Comment by kaikreuzer** [comment]
Not sure how the failing tests could potentially be related to this PR...?

**Comment by hkuhn42** [comment]
I will try to have a look into it tomorrow

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * This service provides functionality around voice services and is the central service to be used directly by others.
+ *
+ * @author Kai Kreuzer - Initial contribution and API
+ */
+public class VoiceManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(VoiceManager.class);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: There is only one static method but that method does not use the logger. We could use a non-static logger

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+    protected void activate(Map<String, Object> config) {
+        modified(config);
+    }
+
+    protected void deactivate() {
+    }
+
+    protected void modified(Map<String, Object> config) {
+        if (config != null) {
+            this.keyword = config.containsKey("keyword") ? config.get("keyword").toString() : "Wakeup";
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: Shouldn't we use string constants?

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+        stt = (stt == null) ? getSTT() : stt;
+        tts = (tts == null) ? getTTS() : tts;
+        hli = (hli == null) ? getHLI() : hli;
+        source = (source == null) ? getSource() : source;
+        sink = (sink == null) ? getSink() : sink;
+        locale = (locale == null) ? localeProvider.getLocale() : locale;
+
+        if (ks != null && stt != null && tts != null && hli != null && source != null && sink != null) {
+            DialogProcessor processor = new DialogProcessor(getKS(), getSTT(), getTTS(), getHLI(), getSource(),
+                    getSink(), localeProvider.getLocale(), keyword);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: I don't know if my reading is correct:
* first we set a variable to the return variable of getVariable if it is null
* then we check if all variables differ from null
* then we use get... for every variable

Why calling the get function again?

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+     *         is found
+     */
+    public TTSService getTTS() {
+        TTSService tts = null;
+        if (defaultTTS != null) {
+            tts = ttsServices.get(defaultTTS);
+            if (tts == null) {
+                logger.warn("Default TTS service '{}' not available!", defaultTTS);
+            }
+        } else if (ttsServices.size() > 0) {
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: If we are not interested in the real size but check if there is an element, we should use isEmpty() (or the negative result of that). Size could be more expensive (runtime) then isEmpty

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+     *         is found
+     */
+    public STTService getSTT() {
+        STTService stt = null;
+        if (defaultTTS != null) {
+            stt = sttServices.get(defaultSTT);
+            if (stt == null) {
+                logger.warn("Default STT service '{}' not available!", defaultSTT);
+            }
+        } else if (sttServices.size() > 0) {
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: See above

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java (hunk 4 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/VoiceManager.java]: Okay, this is repeated in all similar methods.

Can't we use a generic utility function if the stuff is the same for all that functions.

File: bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/DialogProcessor.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * An instance of this class can handle a complete dialog with the user. It orchestrates the keyword spotting, the stt
+ * and tts services together with the human language interpreter.
+ *
+ * @author Kai Kreuzer - Initial contribution and API
+ */
+public class DialogProcessor implements KSListener, STTListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(DialogProcessor.class);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/internal/DialogProcessor.java]: We could rely on a non-static logger.
I will not repeat this for every one for now.

**Comment by kaikreuzer** [comment]
thanks @maggu2810 I have updated the PR.

**Comment by kaikreuzer** [comment]
Tests are green now.
@hkuhn42 Any further comments from your end? If not, I would go ahead and create a CQ for it. We can still continue to work on the code after the merge, my main intention of this PR was to capture the code contributions from the others (before their CLAs mit not be valid anymore or what else could potentially happen).

**Comment by hkuhn42** [comment]
@kaikreuzer no more immediate comments. Please create the cq. I will need a bit more time to have an in depth look as i have to setup a new development environment first.

**Comment by kaikreuzer** [comment]
Thanks for the feedback, @hkuhn42. I have now created CQ 11705 and will wait for its approval.

**Comment by kaikreuzer** [comment]
CQ is approved, hence I merge this now.

**Comment by tavalin** [comment]
@kaikreuzer any hints on where to start using the HLI interfaces? I found what looks to be the reference implementation (StandardInterpreter) but could not use it in my test bundle as it's not an Export-Package.  Then I tried to create my own class that implements HumanLanguageInterpreter but despite adding the Import-Package I received an "unresolved reference" as it seems to be trying to find version "0.0.0" of the package.

**Comment by kaikreuzer** [comment]
Well, [here](https://github.com/eclipse/smarthome/pull/1882/files#diff-77eda7f5c5232fa5fe796bb96ebf3104R25) the package is exported, so it should be available to you. In case you are using the openHAB 2 IDE, please note that I just release an ESH build that includes this PR tonight, so before it wasn't part of the target platform of openHAB.

**Pull request merged**